### ISSUE TITLE
MIR drop order is sometimes odd

### ISSUE THREAD

**Issue by nikomatsakis** [rust-lang/rust#32433]
As @eddyb noted on IRC:
```
15:04 <eddyb> this is weird: http://is.gd/Zjyjmd - it's dropping the outermost temporary, then the innermost, then the middle - shouldn't it be inside-out? or outside-in?
15:04 <eddyb> old trans appears to call them outside-in
15:04 <eddyb> and this drops x before foo(&x) http://is.gd/P1ergF
```

Related test:

src/test/run-pass/issue-23338-ensure-param-drop-order.rs

**Comment by nagisa** [comment]
I’m certainly surprised we do not drop things in reverse order for the 2nd test case. Once we “fix” that, the first issue should fix itself too.

**Comment by eddyb** [comment]
Non-URL-shortener links:
https://play.rust-lang.org/?gist=c20256b92c48181b8218de2897e89b4b&version=nightly&backtrace=0
https://play.rust-lang.org/?gist=0d58d544a41669c6f0c7d6939834ae46&version=nightly&backtrace=0

**Comment by eddyb** [comment]
Better tests (comment out `#[rustc_mir]` to see the difference):
https://play.rust-lang.org/?gist=f0370eec3e8c6404483c8392b3bba601&version=nightly&backtrace=0
https://play.rust-lang.org/?gist=fc8c711fae5f3dd3f8fb3542692d8ac8&version=nightly&backtrace=0

**Comment by pnkfelix** [comment]
Another interesting difference:

```rust
    foo(&foo(&Foo(0), 1), 2);
```
prints: `2 0 1`

while
```rust
    &foo(&foo(&Foo(0), 1), 2);
```
prints: `0 1 2`

**Comment by eddyb** [comment]
@pnkfelix I think that's some statement handling, i.e. the outermost `foo` call in the first example doesn't get the drop from `as_temp`.

**Comment by eddyb** [comment]
Here's my latest testcase: https://play.rust-lang.org/?gist=7da9d41eee66dabeeaa62454f885da6b&version=nightly&backtrace=0
It prints `1 0` correctly (short-lived temporary in tail dropped before variable).
However, MIR, even with the initial fix in #33239, prints `0, 1`.

The problem appears to be that the short-lived `Foo(1)` has `temp_lifetime: Some(CodeExtent(7/Misc(6)))`, where `7` is the ID of `main`'s whole body.
The debug log starts like this:
```
DEBUG:rustc_mir::build::scope: push_scope(CodeExtent(4/CallSiteScope { fn_id: 4, body_id: 6 }))
DEBUG:rustc_mir::build::scope: push_scope(CodeExtent(5/ParameterScope { fn_id: 4, body_id: 6 }))
DEBUG:rustc_mir::build::scope: push_scope(CodeExtent(7/Misc(6)))
DEBUG:rustc_mir::build::scope: push_scope(CodeExtent(8/Remainder(BlockRemainder { block: 6, first_statement_index: 0 })))
```
That means HAIR is telling `rustc_mir::build` that `Foo(1)` *should* only be outlived by the arguments, i.e. the variable should drop *before* `Foo(1)`, and MIR follows suit.

The `CodeExtent` comes from `let temp_lifetime = cx.tcx.region_maps.temporary_scope(expr.id);`.
Does that mean rustc doesn't understand temporaries correctly?

#### Whoopty doo. Days since last soundness bug: 0 (#33490)
